sympy/concrete/zeilberger.py

```diff
@@ -0,0 +1,144 @@
+    we take t_k = a_0F(n, k) + a_1F(n + 1, k) + ... + a_JF(n + J, k) with a_i
+    some variables and determine the gosper normal form of t_(k+1)/t_k.
+    """
+    i = symbols('i', cls = Dummy)
+    a = symbols('a', cls = Function)
+
+    r_1, r_2 = fraction(hypersimp(F, k))
+    s_1, s_2 = fraction(hypersimp(F.subs(n, n-1), n))
+
+    p_0 = Poly(sum(a(j) * (product(s_1.subs(n, n + j - i), (i, 0, j-1)) *
+                    product(s_2.subs(n, n + i), (i, j + 1, J)))
+                    for j in range(J + 1)), k)
+
+    # We can factor out a p_0(k+1)/p_0(k) from the gosper
+    # normal form calculation
+
+    r = r_1 * product(s_2.subs(n, n + i), (i, 1, J))
+    s = r_2 * product(s_2.subs([(n, n + i), (k, k + 1)]), (i, 1, J))
+    r, s = fraction(cancel(r / s))
+
+    p_2, p_3, p_1 = gosper_normal(r, s, k)
+
+    return p_0 * p_1, p_2, p_3.shift(-1)
+
+def zb_recur(F, J, n, k):
+    """
+    Given a hypergeometric function F of n and k (and possibly other variables),
+    determines if there  exists a_0, a_1, ... , a_J polynomials in n and a
+    function G of n and k such that G / F is rational and
+
+    a_0F(n, k) + a_1F(n + 1, k) + ... + a_JF(n + j, k) = G(n, k + 1) - G(n, k).
+
+    If such things exists this function returns a pair consisting of
+    [a_0, a_1, ... , a_J] and the function G / F, otherwise it returns
+    'try higher order', suggesting you try to find a higher order recurrence,
+    i.e. increase J. If F is 'proper hypergeometric' (see page 64 of 'A = B'
+    for a definition) then such a recurrence always exists for some J.
+
+    The main application of this is to sum k over some suitable values so that
+    the G telescopes and you obtain a recurrence for the sum of F(n, k).
+
+    Examples
+    ========
+    Helping us discover that sum of binomial(n - k - 1, k) over k is fibonacci
+
+    >>> from sympy.concrete.zeilberger import zb_recur
+    >>> from sympy.abc import n, k, x
+    >>> from sympy import binomial
+    >>> F = binomial(n - k - 1, k)
+    >>> zb_recur(F, 1, n, k)
+    'try higher order'
+    >>> zb_recur(F, 2, n, k)
+    ([-1, -1, 1], k*(k - n)/((2*k - n)*(2*k - n - 1)))
+
+    Discovering a recurrence for binomial(n, k)**3, the resulting recurrence for
+    the sum can't be solved nicely, we don't care to look at G.
+
+    >>> F = binomial(n, k)**3
+    >>> zb_recur(F, 2, n, k)[0]
+    [-2, -(7*n**2 + 21*n + 16)/(4*(n + 1)**2), (n + 2)**2/(4*(n + 1)**2)]
+
+    F may have other symbols than n and k, here we discover the sum from k = 0 to n to be
+    x / (x + n).
+
+    >>> F = (-1)**k * binomial(n, k) / binomial(x + k, k)
+    >>> zb_recur(F, 1, n, k)
+    ([-n - x, n + x + 1], k*(k + x)/(k - n - 1))
+    """
+    a, w = symbols('a, w', cls = Function)
+
+    p, p_2, p_3 = _zb_gosper(F, J, n, k)
+    # This provides us the polynomials appearing in (6.3.11) of 'A = B', p_3
+    # here is p_3(k-1)
+
+    b_deg = _find_b_deg(p_2, p_3, p)
+
+    if b_deg < 0:
+        return "try higher order"
```

Comment:
Functions should not return strings. If there is an error, it should raise an exception (like `raise ValueError("try higher order")`).